Support autograd.Function w/ grad

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99483

This PR adds support for tracing autograd.Function with grad.

A few important bullet points outlining our approach:

1) Our goal is to verify soundness in order to add a call_function to the autograd.Function's `apply` to the graph.
2) We achieve (1) by either verifying soundness or rejecting soundness, by ensuring that both forward and backward of the autograd.Function are sound.
3) For the forward, if we verify soundness, we install its guards into the graph.
4) For the backward, if we verify soundness, we throw it out. However, backwards soundness verification is more onerous, and has a config driven set of banned attrs and methods for tensors.

1-4 above are achieved by turning the forward and backward into UserDefinedFunctionVariables, and inlining through them, relying on dynamo's soundness detection. If we graph break in these, we raise and treat them as unsound. As noted above, backwards is stricter yet.

For the tracing, the safety comes from dynamo's HigherOrderOperator system. That system ensures that not only do we trace soundly, but that no new variables are lifted into inputs during the tracing, and that the forward and backwards are entirely self contained. 

Whenever we reject a function as unsound, we restore back, as usual. 

Due to some limitations in the lifting logic, we have an escape hatch we implemented for tensors that are known in forward, but cross into backwards through save_tensors (save) /saved_tensors (load). We escape hatch here to avoid having the known saved tensors coming from forward end up being accidentally treated as lifted variables (and rejected). This is sound, but a little hacky feeling.

Additionally, due to some limitations in fx node removal, combined with how we produce subgraphs for the traces installed from HigherOrderOperators, we had to improve our node removal logic. In the event of a restore, we remove the old nodes from the graph, as usual in dynamo. However, because the references to these nodes may exist in subgraphs, we traverse any nodes users and remove them first if and only if they are in another graph. This is always sound, because removal should only be downstream of restoration at this point. 
 


cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
[Experimental] [Needs more text] Rewrite autograd function w/ grad as a TorchVariable and simulate user invoked allow ingraph, 2x faster Deberta training

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Rewrite autograd function w/ grad as a TorchVariable and simulate user invoked allow_in_graph, 2x faster Deberta training"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
This is known to not be sound - you can't just allow_in_graph any old thing like this. However, this (1) will help us measure the potential impact (2) be a base for figuring out how to determine soundness, or, failing that, config it out.

albanD commented on torch/_dynamo/variables/misc.py:
```diff
@@ -224,8 +224,17 @@ def visit(node):
+            def trampoline_autograd_fn(*args, **kwargs):
+                return self.fn_cls.apply(*args, **kwargs)
+
+            torch._dynamo.allow_in_graph(trampoline_autograd_fn)
```
Any way we could get dynamo to trace self._fn_cls.forward to ensure the correctness even if the graph that is passed down contains this call directly?

voznesenskym replied:
Yes, I have plans to ensure this is safe via export fwd and bwd and if it exports, installing the guards from it and proceeding. Still designing the plan :)

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Rewrite autograd function w/ grad as a TorchVariable and simulate user invoked allow_in_graph, 2x faster Deberta training"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Rewrite autograd function w/ grad as a TorchVariable and simulate user invoked allow_in_graph, 2x faster Deberta training"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -224,8 +224,16 @@ def visit(node):
-            unimplemented("autograd.Function with requires_grad")
+            from .torch import TorchHigherOrderOperator
+
+            def trampoline_autograd_fn(*args, **kwargs):
```
This is basically a bad version of CustomFunctionHigherOrderOperator - I did not realize CustomFunctionHigherOrderOperator existed when I wrote this (Thanks @zou3519)

zou3519 replied:
That's okay, we can change it later

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Rewrite autograd function w/ grad as a TorchVariable and simulate user invoked allow_in_graph, 2x faster Deberta training"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented on benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_dynamic_training.csv:
```diff
@@ -8,8 +8,8 @@ BertForQuestionAnswering,pass,7
 CamemBert,pass,7
-DebertaForMaskedLM,pass,52
-DebertaForQuestionAnswering,pass,52
+DebertaForMaskedLM,pass,7
+DebertaForQuestionAnswering,pass,7
```
noob question, but what are these numbers?

voznesenskym replied:
Graph breaks!

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -561,7 +561,7 @@ def inline_user_function_return(self, fn, args, kwargs):
             self.output.guards.update(fn.guards)
             return result
-        except Exception:
+        except Exception as e:
```
nit: don't need to name the exception

voznesenskym replied:
good catch, I was dumping it earlier.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -918,19 +918,25 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
                     # One argument to graph per sub_args
                     for a in sub_args:
-                        if isinstance(a, TensorVariable):
+                        if isinstance(a, ConstantVariable):
+                            proxy = tracer.create_graph_input("const")
```
For my understanding, how does one obtain a ConstantVariable?

voznesenskym replied:
As in, where do they come from? Or mechanically?

Mechanically:

`ConstantVariable(foo)` will do it - foo can be whatever basically. 

They do not have a proxy, though I've thought about maybe doing something akin to ProxiedConstant for when we flip between Constant/SymNode (SymNodeVariable is the closest we have to something like that)

Where they come from:

Seeing a constant during symbolic evaluation.

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -224,8 +224,16 @@ def visit(node):
+            def trampoline_autograd_fn(*args, **kwargs):
+                return self.fn_cls.apply(*args, **kwargs)
+
+            # Speculate fwd
+            # TODO(voz): Check bwd soundness, or something, I dunno, bug Horace
```
We should probably check bwd soundness, or hide this PR behind a config. Otherwise people are going to come to us with silent correctness errors when using autograd.Function with dynamo

zou3519 replied:
bwd soundness is another can of worms though. As @bdhirsh mentions often, we don't really know if the grad_output pass to the backward are Tensors (or None), we don't know their sizes or strides, we don't know if they're Tensor subclasses, etc.

voznesenskym replied:
Yes, I am not sure what to do about this - open to suggestions on how to do it. 

I suppose we could do something like speculate_graph on bwd, but throw out the results?

ngimel replied:
Yeah we could assume  we get all the inputs, and if that's not the case fallback to backward fn?

zou3519 replied:
Here's a lightweight proposal from the comments above:
- follow what AOTAutograd does for grad_outputs. Last time I checked this was to assume the grad_outputs have the same features (size/stride/etc) as the outputs, and it adds coercion of the grad_outputs to the assumed size/stride/etc, but this might be out of date
- speculate_subgraph on bwd on grad_outputs created to be the same size/stride/etc, throw out the results after we've proven safety.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1177,6 +1183,31 @@ def speculate_branch(branch):
+                graph_checkpoint,
+                checkpoint,
+            )
+            p_args = (
+                *(arg.as_proxy() for arg in args),
+                *(arg for arg in body_lifted_freevars),
+            )
```
It looks like we will lift freevars to being inputs of `trampoline_autograd_fn`. Later on, `trampoline_autograd_fn(*args, **kwargs)` will call `autograd_function.apply(*args, **kwargs)`

We haven't actually re-written the autograd.Function to accept more inputs, so this seems wrong. We should probably just graph break if we see free variables (or figure out how to re-install the free variables in trampoline_autograd_fn)

voznesenskym replied:
Good point.

zou3519 reviewed (commented):
This is pretty cool! My take is we should ensure the backward is safe and we may need to be a bit careful about free variables

zou3519 committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
Wow it passed CI

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +234,44 @@ def visit(node):
+            speculated_fwd_result = higher_order_autograd_fn.call_function(
+                tx, args, kwargs
+            )
+            bwd_args = [ctx, speculated_fwd_result]
```
To paraphrase @ezyang - "What about varying strides of grads?". This is an edge case we don't handle here today, but seems moderately safe to ignore for now? Esp since we don't use the result of backward for anything?

zou3519 replied:
It does matter if the autograd.Function branches on the strides of the Tensor and is actually unsafe in one of the branches (maybe there is a print statement or something because the user is debugging their code). The actual case where this matters is if a Tensor is a Tensor subclass -- it would be more likely that in the Tensor subclass case there will be unsafe behavior.

@Chillee's opinion is that this is unlikely to happen.

If we want to be conservative and make this safer, then... we could rewrite the backward function to coerce the grads to be the correct strides (or just speculate contiguous grads and then force the grads to be contiguous), and also have it assert that all the grads are plain old Tensors. We can do this by setting up a new autograd.Function that subclasses from the user's.

Happy to defer to the crowd on how much safety we need here.

voznesenskym replied:
I am going to leave a comment and then punt on this.

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -738,6 +738,86 @@ def handle_ntuple(value):
+# See NOTE [HigherOrderOperator tracing design] for details of the design
+# See NOTE [speculate_subgraph vs old_speculate_subgraph] for other info
+def speculate_subgraph(
+    tx, f, sub_args, graph_checkpoint, checkpoint, *, always_restore=False
```
Is always_restore the worst named variable ever? Maybe we should flip this to speculate_only? or refactor into 2 fns of speculation (arg building + invocation + asserts) from installation onto tx?

zou3519 replied:
`always_restore` or `speculate_only` sound fine to me.

speculate_subgraph was originally rewritten in the second style (see old_speculate_subgraph): it would do "arg building + invocation + asserts", and then the control flow op (cond, map) ends up installing the guards and/or side effects. But it was difficult to get a clean split, because as a part of speculating the subgraph, we actually modify the original graph with things like new graph inputs, so it is now in the form that you see.

voznesenskym commented:
![image](https://user-images.githubusercontent.com/4755252/236016477-c5a0e908-903d-486d-84b6-a366b838205b.png)

albanD commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +234,44 @@ def visit(node):
+            ):
+                unimplemented("Unsafe bwd in autograd.function")
+
+            # If fwd and backward are sound, we want apply in the graph.
```
Not sure about this comment: we want apply in the graph because we need to preserve the autograd behavior for AOTAutograd to be able to do its job.

voznesenskym replied:
Right, I think we are saying the same thing! We *must* have apply in the graph, but we need to check the fwd and bwd to ensure it is sound.

albanD replied:
Yes we agree on what needs to be done (and is done).

> because doing so messes with the versioning of tensors for grad in bwd.

That is the part that is confusing to me as a justification.

voznesenskym replied:
Ah, agreed - if we just write forward in the graph, you don't get all of apply, and therefore, no backward. This is just me musing as I learn how it works, its a low value comment.

albanD commented:
Autograd side sounds good. Definitely very close to the high order operator discussions we were having with @zou3519

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -738,6 +738,86 @@ def handle_ntuple(value):
+            graph = tx.output.graph
+            lifted_freevars = tracer.lifted_freevars
+
+            if always_restore:
+                tx.output.graph = graph_checkpoint
+                tx.restore_graphstate(checkpoint)
```
This never gets hit right? There's already an already_restore above

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -738,6 +738,86 @@ def handle_ntuple(value):
+                assert not isinstance(
+                    a, torch.Tensor
+                ), "Tensors should already be tracked?"
+                if isinstance(a, ConstantVariable):
+                    proxy = tracer.create_graph_input("const")
```
Are ConstVariable always Tensors or can they be other things?

voznesenskym replied:
They should never be tensors, they are things like strings, ints, etc.

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -260,6 +293,16 @@ def call_function(self, tx, args, kwargs):
+    def save_for_backward(self, *tensors: torch.Tensor):
+        super().save_for_backward(tensors)
+        self.saved_tensors.extend([*tensors])
+
```
Not sure if this is handled elsewhere, but people are going to call other methods on the ctx object. If they do this, then want to either raise Unsupported (safe default) or model them as well

voznesenskym replied:
Hmm, well, what we need to model here is what apply cares about, and is not already handled by python. As this is a subclass, we can allow most behavior just fine, I would think? We just need to model anything that gets applied between forward and backward, but not within forward and backward, if that makes sense?

ngimel replied:
Here's a list of context methods that are called reasonably often https://pytorch.org/docs/stable/autograd.html#context-method-mixins, it's fine to not support them and graph break instead, but it's not fine to error out.
These methods can only be applied during forward, not between forward and backward.

voznesenskym replied:
If they are applied during forward, this will already be handled during tracing time by tracing fwd and installing their guards. During the actual runtime, this is handled by apply.

zou3519 replied:
> Answered elsewhere, but maybe we can keep it as is? As long as it stays entire in python in fwd/bwd, it should be fine. save_tensors was special because it gets it's logic done between fwd and bwd, in apply. Alternatively, we could rewrite .apply to do saved_tensors within the context of forward maybe?

Oh I see. Are you saying that when we are speculating the forward, when dynamo sees e.g. ctx.mark_dirty, it will add some guard and then move on?

If that is the case, then yeah, I think we can keep it as is.

zou3519 replied:
> this will already be handled during tracing time by tracing fwd and installing their guards. 

So I tried this, and it looks like this doesn't add any guards. But I think this behavior is correct, since the user has no way to change methods on `ctx` (https://gist.github.com/zou3519/1e90d815ad54ca93efbb0a1e97ecbc23), so we don't actually need to add guards that the method is the method we expect. Does that sounds right?

> If that is the case, then yeah, I think we can keep it as is.

Edge case: We probably want to raise Unsupported if the user does ctx.set_materialize_grads(False). Because this will turn any zero grad_output to the backward into a None, and during introspection we explicitly assume that the backwards grad_output are all Tensors.

Everything else, I agree we can keep as-is

voznesenskym replied:
> during introspection we explicitly assume that the backwards grad_output are all Tensors.

We don't do this for backwards! This is a case we hit in Deberta already, even without set_materialize_grads:
```
    @staticmethod
    def backward(ctx, grad_output):
        if ctx.scale > 1:
            (mask,) = ctx.saved_tensors
            return grad_output.masked_fill(mask, 0) * ctx.scale, None
        else:
            return grad_output, None
```

And we handle this here:
```
            if not isinstance(output, TensorVariable) and not always_restore:
                unimplemented(
                    "HigherOrderOperator with body with non single Tensor output"
                )
```

The idea being, if you soundly got an output, we do not care that it is all tensors or not, since we do not write anything new to the graph here. So this should be safe as well.

voznesenskym replied:
Oh, I understand. Let me handle this case.

voznesenskym replied:
Added a test

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1179,6 +1211,43 @@ def speculate_branch(branch):
+            "trampoline_autograd_bwd",
+            "trampoline_autograd_apply",
+        ):
+            always_restore = self.value.__name__ == "trampoline_autograd_bwd"
```
Hmm, is the reason why we don't `always_restore=True` on trampoline_autograd_fwd is that we need to keep around the outputs so that we can speculate the backward?

voznesenskym replied:
And because we want to install guards from it.

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +234,44 @@ def visit(node):
+            )
+            bwd_args = [ctx, speculated_fwd_result]
+            # ctx.value.saved_tensors = ctx.value.to_save
+            if not is_fn_safe_to_run(
+                tx, TorchHigherOrderOperator(trampoline_autograd_bwd), bwd_args
+            ):
```
Why is this not just `is_fn_safe_to_run(tx, trampoline_autograd_bwd, bwd_args)`?

voznesenskym replied:
We need to make it an operator so we get the higher order operator behavior on call_function, the one that does the actual speculation. 

We could of course, hide that in is_fn_safe_to_run.

zou3519 replied:
That makes sense. No need to hide unless you think that would be nicer.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -738,6 +738,86 @@ def handle_ntuple(value):
+                ), "Tensors should already be tracked?"
+                if isinstance(a, ConstantVariable):
+                    proxy = tracer.create_graph_input("const")
+                elif isinstance(a, (TensorVariable, AutogradFunctionContextVariable)):
+                    tracer.create_graph_input(a.as_proxy().node.name)
```
This makes me a bit uncomfortable, but I think I can live with it:
- let's say we access ctx.saved_tensors[0]. Will this trigger our free variable check?
- can anything be a proxy, or are proxies supposed to be Tensors?

One way around this is that we can have trampoline_fwd and trampoline_bwd "hide" the ctx object. That is, trampoline_fwd could instead return (output, saved_tensors) and trampoline_bwd could accept (actual_args, saved_tensors), and then inside trampoline_bwd we construct a new ctx object and install saved_tensors on it.

voznesenskym replied:
Hmm, I don't hate that idea, but I worry that this might lead to us having to model more than saved_tensors, whereas now AutogradFunctionContextVariable wrapping a real object does a lot of heavy lifting for us.

voznesenskym replied:
Fixed by lazy

zou3519 reviewed (commented):
This (and your benchmarks) is awesome!

I had some high-level comments:
- side effects: we should probably ban them. speculate_subgraph doesn't do that yet
- there are other methods on the ctx object. We can probably just have them raise Unsupported if caught. These would be great to test.
- Proxy-ifying the context object. This makes me a little uncomfortable, because my mental model was that only Tensors should be Proxies (we do things like try to save FakeTensors on Proxy nodes as well). Maybe this is fine and I'm out of date

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -738,6 +738,86 @@ def handle_ntuple(value):
+    # Snapshot state
+    graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
+    # Will raise if not sound
+    output, graph, lifted_freevars = speculate_subgraph(
+        tx, f, sub_args, graph_checkpoint, checkpoint, always_restore=True
+    )
```
Agreed.

voznesenskym commented:
> This (and your benchmarks) is awesome!
> 
> I had some high-level comments:
> 
> * side effects: we should probably ban them. speculate_subgraph doesn't do that yet

Huge +1.

> * there are other methods on the ctx object. We can probably just have them raise Unsupported if caught. These would be great to test.

Answered elsewhere, but maybe we can keep it as is? As long as it stays entire in python in fwd/bwd, it should be fine. save_tensors was special because it gets it's logic done between fwd and bwd, in apply. Alternatively, we could rewrite .apply to do saved_tensors within the context of forward maybe?

> * Proxy-ifying the context object. This makes me a little uncomfortable, because my mental model was that only Tensors should be Proxies (we do things like try to save FakeTensors on Proxy nodes as well). Maybe this is fine and I'm out of date

We already proxy other things, like symints and certain other variables.

That being said, this was just done so we could (1) propogate it as an input (2) let the proxy creation handle the heavy lifting of naming (so it does not falsely get flagged as a free var/lifted). 

No strong feelings here, let me see if we can avoid the proxy...

ezyang commented:
I'm deferring to @zou3519 for this review

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +234,44 @@ def visit(node):
+            # We do not want forward, because doing so messes with the versioning of tensors for grad in bwd.
+            # And we don't want backwards for the obvious reasons.
+            args = args[1:]
+            return TorchHigherOrderOperator(trampoline_autograd_apply).call_function(
```
Yes

zou3519 replied:
Is it possible for the backward to add more guards? If so, should we do anything with them? (Do backward operators in PyTorch today add new guards or do we assume that cannot happen?)

voznesenskym replied:
No, we don't guard anything from/in backward. Today we assume that backward cannot introduce new guards, and all changes guarded on in forward suffice to cover anything that can happen in backwards. We have a design we call k-tier caching that would ostensibly change this paradigm with the hope of having smaller recompile blast radius, by having more guard caches, closer to the potentially changing code.

zou3519 replied:
Thanks for the context. If we assume the backward cannot introduce guards, could we check if the backward pass does, and if so, add a graph break? I am not sure what can go wrong if the user-provided backward does introduce guards (silent incorrectness?)

ngimel commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +234,44 @@ def visit(node):
+            # TODO(voz): NOTE: This function identity is unguarded, but the odds of someone swapping self.fn_cls from
+            # autograd fn to something else is very low.
+            higher_order_autograd_fn = TorchHigherOrderOperator(trampoline_autograd_fwd)
+            speculated_fwd_result = higher_order_autograd_fn.call_function(
```
where is safety of forward checked?

voznesenskym replied:
In TorchHigherOrderOperator call_function->speculate_subgraph, there are safety checks there.

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -0,0 +1,115 @@
+import torch._dynamo.testing
+
+
+class CustomFunc1(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        return foo + foo
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output
+
```
So there is a second type of autograd.Function that has a `setup_context` staticmethod where the forward doesn't actually accept a ctx object (https://pytorch.org/docs/stable/notes/extending.html#example). We don't need to support it in this PR, but it would be good to test that that does not completely blow up underneath torch.compile (today I would expect it to graph break)

voznesenskym replied:
I can add a test for this :) It should gracefully graph break here.

voznesenskym replied:
Test added.

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +234,63 @@ def visit(node):
-            unimplemented("autograd.Function with requires_grad")
+            # Note - this is the same check used in autograd/function.py, except inverted.
+            # If we want to support functorch transforms here, we will need to enable this.
+            if (
```
Not sure if it's there already but we want an assert that neither `vjp` or `jvp` was re-defined on this class either.
The first one because we only look at .backward instead of doing https://github.com/pytorch/pytorch/blob/7a15e823885ddad1696e507a642fc89029a75009/torch/autograd/function.py#L273 and the second because we just ignore forward AD grads.

albanD commented on test/dynamo/test_autograd_function.py:
```diff
@@ -0,0 +1,182 @@
+    @staticmethod
+    def backward(ctx, grad_out1, grad_out2):
+        if grad_out2 is None:
+            print("grad_out2 is None!")
```
debug print?

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +234,63 @@ def visit(node):
+
+            from .torch import is_fn_safe_to_run, TorchHigherOrderOperator
+
+            def trampoline_autograd_apply(*args, **kwargs):
```
Yea, I'll add a comment.

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot rebase

voznesenskym commented:
@pytorchbot rebase

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2128,12 +2138,13 @@ def check_inlineable(func):
-            unimplemented(
-                f"inline in skipfiles: {func.fn.__qualname__}  | {func.get_name()} {func.get_filename()}"
-            )
+        if "variables/misc.py" not in func.get_filename():
```
gross, move to a non dynamo file

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -735,6 +735,82 @@ def handle_ntuple(value):
+    # Will raise if not sound
+    try:
+        f.call_function(tx, sub_args, {})
+    except Exception as e:
+        raise
```
Nit: Does try-finally (without except) do the same thing?

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +234,70 @@ def visit(node):
+            def trampoline_autograd_bwd(*args, **kwargs):
+                return self.fn_cls.backward(*args, **kwargs)
+
+            # Note - On Tracing autograd.Function w/ grad
```
nit: note syntax is usually "NOTE [On Tracing autograd.Function w/ grad]"

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -259,6 +318,32 @@ def call_function(self, tx, args, kwargs):
+                return tx.inline_call(
+                    tx, UserFunctionVariable(self.fn_cls.backward), args, kwargs
+                )
+        return tx.inline_call(
+            tx, UserFunctionVariable(self.fn_cls.forward), args, kwargs
+        )
```
nit: assert `name == 'forward'`?

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -259,6 +318,32 @@ def call_function(self, tx, args, kwargs):
+        )
+
+
+class SaveSimulatingAutogradFunctionContext(torch.autograd.function.FunctionCtx):
```
We discussed this a bit, but do we still need `SaveSimulatingAutogradFunctionContext`?

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -285,22 +380,34 @@ def call_method(
             return LambdaVariable(
                 lambda *args, **kwargs: self.call_method(tx, name, args, kwargs)
             ).add_options(self)
+        if name == "saved_tensors":
+            return self.call_method(tx, name, [], {})
```
nit: Ack, morally `saved_tensors` isn't a method on the ctx object (it's an attribute, right?) Is there a way around this, or is the claim that attributes are just special cases of methods?

voznesenskym replied:
This just uses the call_method logic for saved_tensors, I suppose I could move this down.

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -285,22 +380,34 @@ def call_method(
+            self._saved_tensors = []
+        for arg in args:
+            # as_proxy can return constant values or other non proxy values
+            if isinstance(arg.as_proxy(), torch.fx.Proxy):
+                arg.as_proxy().node.meta["whitelisted"] = True
+            self._saved_tensors.append(arg)
```
Do we need to whitelist things saved directly to the ctx object? Or is that immediately a graph break during the forward?

voznesenskym replied:
Sorry, what do you mean?

zou3519 replied:
If we have an autograd.Function that looks like:
```
class Func(autograd.Function):
    @staticmethod
    def forward(ctx, x):
        ctx.x = x  # Well, we don't recommend users save tensors directly
        return x.view(-1)
    @staticmethod
    def backward(ctx, grads)
        return grads.view(ctx.x.shape)
        
class Func(autograd.Function):
    @staticmethod
    def forward(ctx, x):
        ctx.x_shape = x.shape
        return x.view(-1)
    @staticmethod
    def backward(ctx, grads)
        return grads.view(ctx.x_shape)

```

Do we need to whitelist `ctx.x_shape` or `ctx.x`? `ctx.x_shape` is not a Tensor, so it probably doesn't get a Proxy?

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -229,12 +229,17 @@ def bind_args(self, parent, args, kwargs):
-                        closure_cell_contents = AttrSource(
-                            closure_cell, "cell_contents"
-                        )
+                        if "trampoline_autograd" in self.fn.__name__:
```
this is not very robust, right? It looks like we are in UserFunctionVariable -- a user could write a function named `trampoline_autograd_blah` and it could trigger this

voznesenskym replied:
Yes, I think the way to do this will be direct fn compare once I pull it out into its own toplevel file, instead of being in this nested garbage location that also makes skipfiles gross.

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -229,12 +229,17 @@ def bind_args(self, parent, args, kwargs):
-                            closure_cell, "cell_contents"
-                        )
+                        if "trampoline_autograd" in self.fn.__name__:
+                            closure_cell_contents = self.source
```
Er, what is going on here?

zou3519 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -263,6 +263,9 @@ def call_method(
         kwargs: "Dict[str, VariableTracker]",
     ) -> "VariableTracker":
+        if tx.strict_checks_enabled:
+            if name in ["stride"]:
+                unimplemented(f"Illegal method invocation {name} in strict mode")
```
We should probably also disable all Tensor properties besides size, because those are unknown.

Off of the top of my head: storage_offset, requires_grad (the attribute), dtype, device, layout, all the `is_*` functions (e.g. is_cuda, is_quantized, ...).

Ideally we'd also have a check for `isinstance(x, <something that is not a torch.Tensor>)`, but that is also unlikely in practice.

voznesenskym replied:
We can't get here without it being a tensor, but yes, agreed with the rest.

zou3519 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -0,0 +1,255 @@
+    def backward(ctx, grad_output):
+        (result,) = ctx.saved_tensors
+        return grad_output * math.sqrt(result.numel())
+
```
nit: it's a bit difficult to scroll up and down when reading the tests (because the tests references the modules), but it's not the worst

zou3519 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -0,0 +1,255 @@
+            for model in [Module5(), Module6()]:
+                torch._dynamo.reset()
+                cnts = torch._dynamo.testing.CompileCounter()
+                opt_model = torch._dynamo.optimize(cnts)(model)
```
nit: are all of the torch._dynamo.optimize calls just `torch.compile`?

voznesenskym replied:
Yes but I prefer this.

zou3519 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -0,0 +1,255 @@
+                self.assertEqual(cnts.frame_count, 2)
+
+    def test_linear_setup_context(self):
+        model = ModuleLinear()
```
Is there a reason why we wrap the autograd.Function in an nn.Module instead of having a function invoke the autograd.Function and passing that function to torch.compile?

voznesenskym replied:
This was the prior art in the tests I copied from, and I figured why not keep the pattern.

zou3519 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -0,0 +1,255 @@
+
+    def test_linear_setup_context(self):
+        model = ModuleLinear()
+        opt_model = torch._dynamo.optimize("eager")(model)
+        input = torch.randn(2, 2, dtype=torch.double, requires_grad=True)
+        weight = torch.randn(3, 2, dtype=torch.double, requires_grad=True)
+        opt_model(input, weight)
```
Let's check for graph break and/or that the result is correct? Would be bad if the result is wrong. Ditto for the other tests that are spiritually similar (and are testing that we fallback to eager mode).

Also Alban had a comment somewhere about testing an autograd.Function with a `jvp` staticmethod

voznesenskym replied:
Yes, that is still outstanding.

voznesenskym replied:
Added, and jvp is excluded now.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1179,6 +1207,71 @@ def speculate_branch(branch):
+            post_guards = tx.output.guards
+            if body_lifted_freevars:
+                for freevar in body_lifted_freevars:
+                    if "whitelisted" not in freevar.node.meta:
```
The "whitelisted" thing feels super hacky but I don't have any better suggestions lol

voznesenskym replied:
Yea its a little unfortunate... but not egregious I guess?

zou3519 replied:
also, nit: call it 'allowlisted' or 'came_from_saved_tensors'

voznesenskym replied:
agreed.

zou3519 reviewed (commented):
There are some things I want to read in a bit more detail, but I like the way this is going. Only main item I have is around banning more operations in the safe tracing mode

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[Experimental] [Needs more tests] Support autograd function (2x Deberta!)"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Support autograd function"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD commented on test/test_torch.py:
```diff
@@ -5721,6 +5724,7 @@ def disable_gc():
     else:
         yield
 
+@skipIfTorchInductor("segfaults")
```
This will get fixed before landing right?

albanD reviewed (commented):
Is this ready for review? I see a few debug prints still?
Also it would be great to have a more detailed PR description now that things are more stable!

voznesenskym committed:
Update on "Support autograd function"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +258,75 @@ def visit(node):
+
+            # if both the forward and backward traces are sound, we write the autograd function’s apply into the graph.
+
+            # For tracing forward and backward, we use UserFunctionVariable. Although it does not directly contribute
+            # to soundness evaluation, it plus a  GlobalSource makes sure we can produce valid guards,
+            # and that we can inline properly here. Inlining is required in order to be able to ensure that the
+            # soundness evaluation works as described above.
```
My intuition for why we are using UserFunctionVariable is that the user wrote the autograd.Function's forward and backward. The stuff about inlining makes sense though.

voznesenskym replied:
Yes :)

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +258,75 @@ def visit(node):
+            # graph.We trace forward, and evaluate soundness. Soundness, in this context, refers to the absence of side
+            # effects, the avoidance of lifting new arguments into the trace, the production of a single tensor output,
+            # and a limited input scope confined to contexts, tensors, and constants. If the forward trace is sound,
+            # we install any guards accumulated from tracing. If not, we graph break. We trace backward, and evaluate
+            # for soundness, same as forward, except with more strictness. We enable a strict mode on the tx, and
+            # reject certain ops when running under this strict mode. If the backward trace is sound, we discard the
+            # trace by restoring. Otherwise, we raise.
```
What happens to guards from the backward?

zou3519 replied:
Answered later on: we check if they are the same or not and then raise Unsupported if they aren't.

zou3519 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -256,13 +260,31 @@ def unpack_var_sequence(self, tx, idxes=None):
+            "is_meta",
+            "data",
+            "is_sparse",
+            "ndim",
```
ndim is fine, torch.compile specializes on the number of dimensions and we do know ahead of time how many dimensions the grads have.

If we want to make this more robust we could blanket ban any method that starts with `is_`, because if you're querying this you are probably about to switch on it.

voznesenskym replied:
ack

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -292,15 +407,23 @@ def call_method(
             return LambdaVariable(
                 lambda *args, **kwargs: self.call_method(tx, name, args, kwargs)
             ).add_options(self)
+        if name == "saved_tensors":
+            return variables.TupleVariable(list(self._saved_tensors))
```
nit: should send a tuple to a TupleVariable instead of list?

voznesenskym replied:
I think it expects `items` to be a list. Grep for `TupleVariable(list` shows the pattern :P

zou3519 replied:
Lol wat

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +258,75 @@ def visit(node):
+            trampoline_autograd_fwd = produce_trampoline_autograd_fwd(self.fn_cls)
+            trampoline_autograd_bwd = produce_trampoline_autograd_bwd(self.fn_cls)
+
+            # NOTE [On Tracing autograd.Function w/ grad]
```
Though, to Alban's point, some elaboration on the design might be helpful for the future. e.g. for handling speculation of a function with a ctx object, we have turned ctx in a proxy and whitelisted ctx.saved_tensors from the free variable lifting mechanism.

voznesenskym replied:
> Left some comments and minor nits, the two points I want to make sure of is:
> 
> * safety: that we're disallowing all the methods/attrs we should be for a safe backwards trace

Yea, this is tricky, because we want to balance restrictiveness with perf because we are being overcautious. The attrs are not necessarily bad, but control flow on them is. I might make this a config driven list.

> * question about the testing plan: I remember the set_materialize_grads thing ended up exposing a bug where Dynamo would straight up yell instead of gracefully fallback to eager. It would be bad if we shipped this PR and then regressed existing user code that do have autograd.Function in them esp because autograd.Function is a very popular API. Do we have a good set of test cases for autograd.Function? (like, in torchbench?)

I don't know! More tests is better though, I am happy to add more tests. Fwiw - benchmark models have autograd.Function in them, and they pass, which is a good representation of reality. 

>   * Related: is it worth adding a dynamo config flag to turn off this feature? If config is True, then always graph break on autograd.Function. Or maybe that is just torch._dynamo.disallow_in_graph, or torch._dynamo.disable?

I can add a kill flag for this, on by default. 

> Otherwise, code LGTM

Thanks, responded to comments.

zou3519 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -256,13 +260,31 @@ def unpack_var_sequence(self, tx, idxes=None):
 
+    def _strict_mode_banned_ops(self):
+        return [
+            "stride",
```
dtype should be here, though you did mention that it did cause a graph break in the model you were looking at. We should figure out if we want to include it or not

voznesenskym replied:
> Left some comments and minor nits, the two points I want to make sure of is:
> 
> * safety: that we're disallowing all the methods/attrs we should be for a safe backwards trace

Yea, this is tricky, because we want to balance restrictiveness with perf because we are being overcautious. The attrs are not necessarily bad, but control flow on them is. I might make this a config driven list.

> * question about the testing plan: I remember the set_materialize_grads thing ended up exposing a bug where Dynamo would straight up yell instead of gracefully fallback to eager. It would be bad if we shipped this PR and then regressed existing user code that do have autograd.Function in them esp because autograd.Function is a very popular API. Do we have a good set of test cases for autograd.Function? (like, in torchbench?)

I don't know! More tests is better though, I am happy to add more tests. Fwiw - benchmark models have autograd.Function in them, and they pass, which is a good representation of reality. 

>   * Related: is it worth adding a dynamo config flag to turn off this feature? If config is True, then always graph break on autograd.Function. Or maybe that is just torch._dynamo.disallow_in_graph, or torch._dynamo.disable?

I can add a kill flag for this, on by default. 

> Otherwise, code LGTM

Thanks, responded to comments.

voznesenskym replied:
Move to config.

zou3519 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -0,0 +1,255 @@
+        return grad_output
+
+
+class CustomFunc2(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        return foo + foo
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output
+
```
Nit: CustomFunc1 and CustomFunc2 are the same thing?

zou3519 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -0,0 +1,255 @@
+        return grad_output * math.sqrt(result.numel())
+
+
+class Module1(torch.nn.Module):
+    def forward(self, foo):
+        return CustomFunc1().apply(foo)
+
+
+class Module2(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.fn = CustomFunc1.apply
+
+    def forward(self, foo):
+        return self.fn(foo)
+
+
+class Module3(torch.nn.Module):
+    def forward(self, foo):
+        return CustomFunc2().apply(foo)
+
+
+class Module4(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.fn = CustomFunc2.apply
+
+    def forward(self, foo):
+        return self.fn(foo)
+
```
Modules 1-4 seem unused?

Also, we have almost no small tests that the mechanism actually works (aside from the graph break counts in the CSV and test_repros). We should add some sanity checks for some simple autograd.Function that Dynamo doesn't graph break

voznesenskym replied:
Accessed in a loop test via `model = globals()[f"Module{i}"]()`

zou3519 reviewed (commented):
Left some comments and minor nits, the two points I want to make sure of is:
- safety: that we're disallowing all the methods/attrs we should be for a safe backwards trace
- question about the testing plan: I remember the set_materialize_grads thing ended up exposing a bug where Dynamo would straight up yell instead of gracefully fallback to eager. It would be bad if we shipped this PR and then regressed existing user code that do have autograd.Function in them esp because autograd.Function is a very popular API. Do we have a good set of test cases for autograd.Function? (like, in torchbench?)
   - Related: is it worth adding a dynamo config flag to turn off this feature? If config is True, then always graph break on autograd.Function. Or maybe that is just torch._dynamo.disallow_in_graph, or torch._dynamo.disable?

Otherwise, code LGTM

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -212,6 +212,30 @@ def __init__(self, inspected, **kwargs):
         self.inspected = inspected
 
 
+def produce_trampoline_autograd_fwd(fn_cls):
+    def trampoline_autograd_fwd(*args, **kwargs):
+        return fn_cls.forward(*args, **kwargs)
+
+    trampoline_autograd_fwd._origin = produce_trampoline_autograd_fwd
+    return trampoline_autograd_fwd
+
+
+def produce_trampoline_autograd_bwd(fn_cls):
+    def trampoline_autograd_bwd(*args, **kwargs):
+        return fn_cls.backward(*args, **kwargs)
+
+    trampoline_autograd_bwd._origin = produce_trampoline_autograd_bwd
+    return trampoline_autograd_bwd
+
+
+def produce_trampoline_autograd_apply(fn_cls):
+    def trampoline_autograd_apply(*args, **kwargs):
+        return fn_cls.apply(*args, **kwargs)
+
+    trampoline_autograd_apply._origin = produce_trampoline_autograd_apply
+    return trampoline_autograd_apply
```
Yes.

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +258,75 @@ def visit(node):
+                    "NYI - autograd.Function with custom setup_context method"
+                )
+
+            vjp_fn = self.fn_cls.vjp  # type: ignore[attr-defined]
+            if vjp_fn is not torch.autograd.Function.vjp:
+                unimplemented("NYI - User defind vjp")
```
ack

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +258,75 @@ def visit(node):
+            module_source = AttrSource(
+                tx.import_source(self.fn_cls.__module__), self.fn_cls.__name__
+            )
+            trampoline_autograd_fwd._source = AttrSource(module_source, "forward")
```
Yea, good point, let me lift this into TorchHigherOrderOperator - I initially wasnt sure if we even needed it.

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -259,6 +347,23 @@ def call_function(self, tx, args, kwargs):
+                return tx.inline_call(
+                    tx, UserFunctionVariable(self.fn_cls.backward), args, kwargs
+                )
+        assert name == "forward"
```
This assert will already get caught and turned into unsupported, but we could be more explicit, sure!

zou3519 replied:
Oh I didn't realize. Are all asserts caught and turned into Unsupported?

voznesenskym replied:
Yes, at least in this part of the flow due to how we wrap inlining

voznesenskym commented:
> Left some comments and minor nits, the two points I want to make sure of is:
> 
> * safety: that we're disallowing all the methods/attrs we should be for a safe backwards trace

Yea, this is tricky, because we want to balance restrictiveness with perf because we are being overcautious. The attrs are not necessarily bad, but control flow on them is. I might make this a config driven list.

> * question about the testing plan: I remember the set_materialize_grads thing ended up exposing a bug where Dynamo would straight up yell instead of gracefully fallback to eager. It would be bad if we shipped this PR and then regressed existing user code that do have autograd.Function in them esp because autograd.Function is a very popular API. Do we have a good set of test cases for autograd.Function? (like, in torchbench?)

I don't know! More tests is better though, I am happy to add more tests. Fwiw - benchmark models have autograd.Function in them, and they pass, which is a good representation of reality. 

>   * Related: is it worth adding a dynamo config flag to turn off this feature? If config is True, then always graph break on autograd.Function. Or maybe that is just torch._dynamo.disallow_in_graph, or torch._dynamo.disable?

I can add a kill flag for this, on by default. 

> Otherwise, code LGTM

Thanks, responded to comments.

voznesenskym committed:
Update on "Support autograd function"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/config.py:
```diff
@@ -258,6 +258,20 @@ def is_fbcode():
+
+_autograd_backward_strict_mode_banned_ops = [
+    "stride",
+    "requires_grad",
+    "storage_offset",
+    "layout",
+    "is_cuda",
+    "is_quantized",
+    "is_meta",
+    "data",
+    "is_sparse",
```
Could we do a regex and ban all `is_*` methods? The naming implies those are Tensor properties. Alternatively, we could add all of the ones in https://pytorch.org/docs/stable/tensors.html to the list, but that might not be as robust:
<img width="951" alt="image" src="https://github.com/pytorch/pytorch/assets/5652049/8d94a169-4510-40e1-a8bf-c5c134c33357">

zou3519 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -234,11 +258,77 @@ def visit(node):
+
+            jvp_fn = self.fn_cls.jvp  # type: ignore[attr-defined]
+            if jvp_fn is not torch.autograd.Function.jvp:
+                unimplemented("NYI - User defind jvp")
```
nit: defind -> defined

zou3519 reviewed (approved):
I convinced myself that `dtype` is not actually a problem. Added some more suggestions for methods to ban.

voznesenskym committed:
Update on "Support autograd.Function w/ grad"


This PR adds support for tracing autograd.Function with grad.

A few important bullet points outlining our approach:

1) Our goal is to verify soundness in order to add a call_function to the autograd.Function's `apply` to the graph.
2) We achieve (1) by either verifying soundness or rejecting soundness, by ensuring that both forward and backward of the autograd.Function are sound.
3) For the forward, if we verify soundness, we install its guards into the graph.
4) For the backward, if we verify soundness, we throw it out. However, backwards soundness verification is more onerous, and has a config driven set of banned attrs and methods for tensors.

1-4 above are achieved by turning the forward and backward into UserDefinedFunctionVariables, and inlining through them, relying on dynamo's soundness detection. If we graph break in these, we raise and treat them as unsound. As noted above, backwards is stricter yet.

For the tracing, the safety comes from dynamo's HigherOrderOperator system. That system ensures that not only do we trace soundly, but that no new variables are lifted into inputs during the tracing, and that the forward and backwards are entirely self contained. 

Whenever we reject a function as unsound, we restore back, as usual. 

Due to some limitations in the lifting logic, we have an escape hatch we implemented for tensors that are known in forward, but cross into backwards through save_tensors (save) /saved_tensors (load). We escape hatch here to avoid having the known saved tensors coming from forward end up being accidentally treated as lifted variables (and rejected). This is sound, but a little hacky feeling.

Additionally, due to some limitations in fx node removal, combined with how we produce subgraphs for the traces installed from HigherOrderOperators, we had to improve our node removal logic. In the event of a restore, we remove the old nodes from the graph, as usual in dynamo. However, because the references to these nodes may exist in subgraphs, we traverse any nodes users and remove them first if and only if they are in another graph. This is always sound, because removal should only be downstream of restoration at this point. 
 


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Support autograd.Function w/ grad"


This PR adds support for tracing autograd.Function with grad.

A few important bullet points outlining our approach:

1) Our goal is to verify soundness in order to add a call_function to the autograd.Function's `apply` to the graph.
2) We achieve (1) by either verifying soundness or rejecting soundness, by ensuring that both forward and backward of the autograd.Function are sound.
3) For the forward, if we verify soundness, we install its guards into the graph.
4) For the backward, if we verify soundness, we throw it out. However, backwards soundness verification is more onerous, and has a config driven set of banned attrs and methods for tensors.

1-4 above are achieved by turning the forward and backward into UserDefinedFunctionVariables, and inlining through them, relying on dynamo's soundness detection. If we graph break in these, we raise and treat them as unsound. As noted above, backwards is stricter yet.

For the tracing, the safety comes from dynamo's HigherOrderOperator system. That system ensures that not only do we trace soundly, but that no new variables are lifted into inputs during the tracing, and that the forward and backwards are entirely self contained. 

Whenever we reject a function as unsound, we restore back, as usual. 

Due to some limitations in the lifting logic, we have an escape hatch we implemented for tensors that are known in forward, but cross into backwards through save_tensors (save) /saved_tensors (load). We escape hatch here to avoid having the known saved tensors coming from forward end up being accidentally treated as lifted variables (and rejected). This is sound, but a little hacky feeling.

Additionally, due to some limitations in fx node removal, combined with how we produce subgraphs for the traces installed from HigherOrderOperators, we had to improve our node removal logic. In the event of a restore, we remove the old nodes from the graph, as usual in dynamo. However, because the references to these nodes may exist in subgraphs, we traverse any nodes users and remove them first if and only if they are in another graph. This is always sound, because removal should only be downstream of restoration at this point. 
 


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge -f "Broken base cse test on macos, 180 other test suites passed"

voznesenskym committed:
Update on "Support autograd.Function w/ grad"


This PR adds support for tracing autograd.Function with grad.

A few important bullet points outlining our approach:

1) Our goal is to verify soundness in order to add a call_function to the autograd.Function's `apply` to the graph.
2) We achieve (1) by either verifying soundness or rejecting soundness, by ensuring that both forward and backward of the autograd.Function are sound.
3) For the forward, if we verify soundness, we install its guards into the graph.
4) For the backward, if we verify soundness, we throw it out. However, backwards soundness verification is more onerous, and has a config driven set of banned attrs and methods for tensors.

1-4 above are achieved by turning the forward and backward into UserDefinedFunctionVariables, and inlining through them, relying on dynamo's soundness detection. If we graph break in these, we raise and treat them as unsound. As noted above, backwards is stricter yet.

For the tracing, the safety comes from dynamo's HigherOrderOperator system. That system ensures that not only do we trace soundly, but that no new variables are lifted into inputs during the tracing, and that the forward and backwards are entirely self contained. 

Whenever we reject a function as unsound, we restore back, as usual. 

Due to some limitations in the lifting logic, we have an escape hatch we implemented for tensors that are known in forward, but cross into backwards through save_tensors (save) /saved_tensors (load). We escape hatch here to avoid having the known saved tensors coming from forward end up being accidentally treated as lifted variables (and rejected). This is sound, but a little hacky feeling.

Additionally, due to some limitations in fx node removal, combined with how we produce subgraphs for the traces installed from HigherOrderOperators, we had to improve our node removal logic. In the event of a restore, we remove the old nodes from the graph, as usual in dynamo. However, because the references to these nodes may exist in subgraphs, we traverse any nodes users and remove them first if and only if they are in another graph. This is always sound, because removal should only be downstream of restoration at this point. 
 


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Support autograd.Function w/ grad"


This PR adds support for tracing autograd.Function with grad.

A few important bullet points outlining our approach:

1) Our goal is to verify soundness in order to add a call_function to the autograd.Function's `apply` to the graph.
2) We achieve (1) by either verifying soundness or rejecting soundness, by ensuring that both forward and backward of the autograd.Function are sound.
3) For the forward, if we verify soundness, we install its guards into the graph.
4) For the backward, if we verify soundness, we throw it out. However, backwards soundness verification is more onerous, and has a config driven set of banned attrs and methods for tensors.

1-4 above are achieved by turning the forward and backward into UserDefinedFunctionVariables, and inlining through them, relying on dynamo's soundness detection. If we graph break in these, we raise and treat them as unsound. As noted above, backwards is stricter yet.

For the tracing, the safety comes from dynamo's HigherOrderOperator system. That system ensures that not only do we trace soundly, but that no new variables are lifted into inputs during the tracing, and that the forward and backwards are entirely self contained. 

Whenever we reject a function as unsound, we restore back, as usual. 

Due to some limitations in the lifting logic, we have an escape hatch we implemented for tensors that are known in forward, but cross into backwards through save_tensors (save) /saved_tensors (load). We escape hatch here to avoid having the known saved tensors coming from forward end up being accidentally treated as lifted variables (and rejected). This is sound, but a little hacky feeling.

Additionally, due to some limitations in fx node removal, combined with how we produce subgraphs for the traces installed from HigherOrderOperators, we had to improve our node removal logic. In the event of a restore, we remove the old nodes from the graph, as usual in dynamo. However, because the references to these nodes may exist in subgraphs, we traverse any nodes users and remove them first if and only if they are in another graph. This is always sound, because removal should only be downstream of restoration at this point. 
 


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Support autograd.Function w/ grad"


This PR adds support for tracing autograd.Function with grad.

A few important bullet points outlining our approach:

1) Our goal is to verify soundness in order to add a call_function to the autograd.Function's `apply` to the graph.
2) We achieve (1) by either verifying soundness or rejecting soundness, by ensuring that both forward and backward of the autograd.Function are sound.
3) For the forward, if we verify soundness, we install its guards into the graph.
4) For the backward, if we verify soundness, we throw it out. However, backwards soundness verification is more onerous, and has a config driven set of banned attrs and methods for tensors.

1-4 above are achieved by turning the forward and backward into UserDefinedFunctionVariables, and inlining through them, relying on dynamo's soundness detection. If we graph break in these, we raise and treat them as unsound. As noted above, backwards is stricter yet.

For the tracing, the safety comes from dynamo's HigherOrderOperator system. That system ensures that not only do we trace soundly, but that no new variables are lifted into inputs during the tracing, and that the forward and backwards are entirely self contained. 

Whenever we reject a function as unsound, we restore back, as usual. 

Due to some limitations in the lifting logic, we have an escape hatch we implemented for tensors that are known in forward, but cross into backwards through save_tensors (save) /saved_tensors (load). We escape hatch here to avoid having the known saved tensors coming from forward end up being accidentally treated as lifted variables (and rejected). This is sound, but a little hacky feeling.

Additionally, due to some limitations in fx node removal, combined with how we produce subgraphs for the traces installed from HigherOrderOperators, we had to improve our node removal logic. In the event of a restore, we remove the old nodes from the graph, as usual in dynamo. However, because the references to these nodes may exist in subgraphs, we traverse any nodes users and remove them first if and only if they are in another graph. This is always sound, because removal should only be downstream of restoration at this point. 
 


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge -f "Broken base cse test on macos, 180 other test suites passed"
